### PR TITLE
[Return types #5.2] Fix param shift

### PR DESCRIPTION
**Pull request by rmoyard**
**Description of the Change:**

Remove unnecessary tuple for  tapes with a single parameter.

e.g.
```
qml.enable_return()
import pennylane as qml

params = np.array([0.1, -1.6, np.pi / 5])

with qml.tape.QuantumTape() as tape:
    qml.RX(params[0], wires=[0])
    qml.CNOT(wires=[0, 1])
    qml.RY(-1.6, wires=[0])
    qml.RY(params[1], wires=[1])
    qml.CNOT(wires=[1, 0])
    qml.RX(params[2], wires=[0])
    qml.CNOT(wires=[0, 1])
    qml.expval(qml.PauliZ(0))
    #qml.probs(wires=[0, 1])
    #qml.var(qml.PauliZ(0))

tape.trainable_params = {0}
dev = qml.device("default.qubit", wires=2)

tapes, fn = qml.gradients.param_shift(tape)
res = fn(dev.batch_execute(tapes))
print(res)
print(type(res))
```

-0.5845993965814567
<class 'numpy.ndarray'>

instead of (np.array(-0.5845993965814567), )

Further changes to the parameter-shift transform:
* The new return types version of the transform is now separated into the `_param_shift_new` function - the number of places where a change in logic was required grew and it made sense to create such a func;
* `_param_shift_new` depends on some edge cases handled by the finite-diff transform;
* Auxiliary functions have been updated including some comments, object names and docstrings.

### CODE REVIEW

File: pennylane/gradients/parameter_shift.py [hunk 1]
```diff
@@ -558,115 +567,124 @@ def _process_pda2_involutory(tape, pdA2, var_idx, non_involutory):
...
             determine the number of results to post-process later
-        non_involutory (list): the indices in the measurement queue of all non-involutory
+        non_involutory_indices (list): the indices in the measurement queue of all non-involutory
             observables
     """
 
     def processing_fn(results):
         # We need to expand the dimensions of the variance mask,
         # and convert it to be the same type as the results.
-        res = results[0]
```
Comment (antalszava) [inline pennylane/gradients/parameter_shift.py:611]: On another look, this section served no direct purpose with our results being tuples/single array objects.

File: pennylane/gradients/parameter_shift.py [hunk 2]
```diff
@@ -558,115 +567,119 @@ def _process_pda2_involutory(tape, pdA2, var_idx, non_involutory):
...
+                else:
+                    m_res = tuple(pdA[m_idx][p_idx] for p_idx in range(num_params))
                 var_grad.append(m_res)
             return tuple(var_grad)
 
-        # Scalar
-        var_grad = _get_var_with_second_order(pdA2, f0, pdA) if mask else pdA
-        return var_grad
+        # Single measurement case, meas has to be variance
+        if num_params == 1:
```
Comment (rmoyard) [inline pennylane/gradients/parameter_shift.py:672]: Single measurement or single parameter?

Comment (antalszava) [reply pennylane/gradients/parameter_shift.py:672]: Single measurement case, then the `if` statement is specific to the single parameter case. The comment is meant to remind the reader that we've covered the multi-measurement case above.

Comment (rmoyard) [reply pennylane/gradients/parameter_shift.py:672]: Thanks!

File: pennylane/gradients/finite_difference.py [hunk 3]
```diff
@@ -154,6 +154,39 @@ def finite_diff_coeffs(n, approx_order, strategy):
     return coeffs_and_shifts
 
 
+def _no_trainable_grad_new(tape):
```
Comment (antalszava) [inline pennylane/gradients/finite_difference.py:157]: These aux functions are used by the param-shift transform too.

File: pennylane/gradients/parameter_shift.py [hunk 4]
```diff
@@ -558,115 +567,119 @@ def _process_pda2_involutory(tape, pdA2, var_idx, non_involutory):
...
-        if i in var_idx:
-            obs_involutory = i not in non_involutory
-            if obs_involutory:
-                num_params = len(tape.trainable_params)
-                item = tuple(np.array(0) for _ in range(num_params))
+
+        if i in involutory_indices:
+            num_params = len(tape.trainable_params)
+            if num_params == 1:
+                item = qml.math.array(0)
```
Comment (eddddddy) [inline pennylane/gradients/parameter_shift.py:574]: Is it necessary to create an array here? It seems that later we only check equality with 0.

File: pennylane/gradients/parameter_shift.py [hunk 5]
```diff
@@ -927,7 +940,7 @@ def processing_fn(results):
 
 
 @gradient_transform
-def param_shift(
+def _param_shift_new(
```
Comment (eddddddy) [inline pennylane/gradients/parameter_shift.py:943]: Worthwhile to update the examples in this docstring, or at least mention that this function replaces `param_shift` in the new return types system.

**Review by eddddddy — APPROVED** [review]
Thanks for the changes @rmoyard 🔥 🔥 🔥! Looks good from my end.

File: pennylane/gradients/finite_difference.py [hunk 6]
```diff
@@ -208,14 +241,7 @@ def _finite_diff_new(
...
     if argnum is None and not tape.trainable_params:
-        warnings.warn(
-            "Attempted to compute the gradient of a tape with no trainable parameters. "
-            "If this is unintended, please mark trainable parameters in accordance with the "
-            "chosen auto differentiation framework, or via the 'tape.trainable_params' property."
-        )
-        if len(tape.measurements) == 1:
-            return [], lambda _: qml.math.zeros([0])
-        return [], lambda _: tuple(qml.math.zeros([0]) for _ in range(len(tape.measurements)))
+        return _no_trainable_grad_new(tape)
```
Comment (rmoyard) [inline pennylane/gradients/finite_difference.py:244]: Thanks for that!

**Review by rmoyard — COMMENTED** [review]
Thanks @antalszava, amazing new test 🥇  I cannot approve I let you do it.

File: pennylane/gradients/parameter_shift.py (hunk 4 above)
Comment (antalszava) [reply pennylane/gradients/parameter_shift.py:574]: Yes, it is. This is a case where `pdA2` is not `0`, but components of the gradient may be.

File: pennylane/gradients/parameter_shift.py (hunk 5 above)
Comment (antalszava) [reply pennylane/gradients/parameter_shift.py:944]: Added TODO and a comment.

**Review by antalszava — APPROVED** [review]

**Pull request merged**